File: 백준/Gold/1504. 특정한 최단 경로/특정한 최단 경로.cc
#include<bits/stdc++.h>
using namespace std;
using ll = long long;
using pii = pair<int, int>;

const int INF = 1e9;
vector<vector<pii>> e;
int n, m;

ll dij(int start, int end){
    vector<int> dist(n+1, INF);
    dist[start] = 0;
    priority_queue<pii, vector<pii>, greater<pii>> pq; pq.push({0, start});
    while(pq.size()){
        auto [w, cur] = pq.top(); pq.pop();
        if(cur == end) return w;
        if(dist[cur] < w) continue;
        for(auto [nxt, nw]: e[cur]){
            if(dist[nxt] > w+nw){
                dist[nxt] = w+nw;
                pq.push({dist[nxt], nxt});
            }
        }
    }
    return INF;
}

int main(){
    cin.tie(0); cout.tie(0); ios::sync_with_stdio(false);
    cin >> n >> m; e.resize(n+1);
    for(int i = 0; i < m; i++){
        int u, v, w; cin >> u >> v >> w;
        e[u].push_back({v, w});
        e[v].push_back({u, w});
    }
    int v1, v2; cin >> v1 >> v2;
    ll ans = min(dij(1, v1)+dij(v2, n), dij(1, v2)+dij(v1, n))+dij(v1, v2);
    cout << (ans >= INF ? -1 : ans);
    return 0;
} 

File: 백준/Gold/1504. 특정한 최단 경로/특정한 최단 경로.py
import sys 
print = lambda *args, sep=" ", end="\n": sys.stdout.write(sep.join(map(str, args)) + end)
input = lambda: sys.stdin.readline().rstrip('\r\n')

import heapq
from math import inf

def main():
    n, m = map(int, input().split())
    e = [[] for _ in range(n+1)]
    for _ in range(m):
        u, v, w = map(int, input().split())
        e[u].append((v, w))
        e[v].append((u, w))
    v1, v2 = map(int, input().split())
    def dij(start):
        dist = [inf]*(n+1)
        dist[start] = 0
        hq = [(0, start)]
        while hq:
            w, cur = heapq.heappop(hq)
            if dist[cur] < w: continue
            for nxt, nw in e[cur]:
                if dist[nxt] > w+nw:
                    dist[nxt] = w+nw
                    heapq.heappush(hq, (dist[nxt], nxt))           
        return dist
    distv1, distv2 = dij(v1), dij(v2)
    path1 = distv1[1]+distv1[v2]+distv2[n]
    path2 = distv2[1]+distv2[v1]+distv1[n]
    ans = min(path1, path2)
    print(-1 if ans == inf else ans)
    return
if __name__ == '__main__':
    main()